src/nupic/research/temporal_memory.py

```diff
@@ -545,96 +407,77 @@ def leastUsedCell(self, cells, connections):
     return sorted(leastUsedCells)[i]
 
 
-  @staticmethod
-  def activeSynapsesForSegment(segment, activeCells, connections):
+  def growSynapses(self, nDesiredNewSynapes, prevWinnerCells, segment):
     """
-    Returns the synapses on a segment that are active due to lateral input
-    from active cells.
+    @params nDesiredNewSynapes (int)  Desired number of synapses to grow
+    @params prevWinnerCells    (list) Winner cells in `t-1`
+    @param  segment            (int)  Segment to grow synapses on.
 
-    @param segment     (int)         Segment index
-    @param activeCells (set)         Indices of active cells
-    @param connections (Connections) Connectivity of layer
-
-    @return (set) Indices of active synapses on segment
+    Notes: The process of swapping the last value into the index in the array
+    that was most recently changed is to ensure the same results that we get
+    in the c++ implentation using iter_swap with vectors.
     """
-    synapses = set()
+    candidates = set(prevWinnerCells)
+
+    for synapse in self.connections.synapsesForSegment(segment):
+      presynapticCell = self.connections.dataForSynapse(synapse).presynapticCell
 
-    for synapse in connections.synapsesForSegment(segment):
-      synapseData = connections.dataForSynapse(synapse)
+      if presynapticCell in candidates:
+        candidates.remove(presynapticCell)
 
-      if synapseData.presynapticCell in activeCells:
-        synapses.add(synapse)
+    candidates = sorted(candidates)
```

Comment:
This method is definitely fine, but it's basically using two strategies for removing candidates, and I think it would make sense to just use one strategy. Do you think it would be hard to do swapping for the first half of this method, rather than set `remove`? Then it would be like the C++, which does all of this with a single candidates vector, with no extra sorting.